File: elixir/euchre/lib/deck.ex
defmodule Euchre.Deck do
  alias Euchre.Card
  def gen do
     [%Card{suit: :hearts, rank: 8},
      %Card{suit: :spades, rank: 8},
      %Card{suit: :clubs, rank: 8},
      %Card{suit: :diamonds, rank: 8},
      %Card{suit: :hearts, rank: 9},
      %Card{suit: :spades, rank: 9},
      %Card{suit: :clubs, rank: 9},
      %Card{suit: :diamonds, rank: 9},
      %Card{suit: :hearts, rank: 10},
      %Card{suit: :spades, rank: 10},
      %Card{suit: :clubs, rank: 10},
      %Card{suit: :diamonds, rank: 10},
      %Card{suit: :hearts, rank: 'Q'},
      %Card{suit: :spades, rank: 'Q'},
      %Card{suit: :clubs, rank: 'Q'},
      %Card{suit: :diamonds, rank: 'Q'},
      %Card{suit: :hearts, rank: 'K'},
      %Card{suit: :spades, rank: 'K'},
      %Card{suit: :clubs, rank: 'K'},
      %Card{suit: :diamonds, rank: 'K'},
      %Card{suit: :hearts, rank: 'A'},
      %Card{suit: :spades, rank: 'A'},
      %Card{suit: :clubs, rank: 'A'},
      %Card{suit: :diamonds, rank: 'A'},
      %Card{suit: :hearts, rank: 'J'},
      %Card{suit: :spades, rank: 'J'},
      %Card{suit: :clubs, rank: 'J'},
      %Card{suit: :diamonds, rank: 'J'}]
  end

  def init(deck)
    when is_list(deck)
  do
    deck
    |> Enum.shuffle
    |> Enum.slice(3, 28)
    |> turn
  end

  def get_top_card(deck)
    when is_list(deck)
  do
    [head| tail] = deck
    %{head: head, deck: tail}
  end

  defp turn(deck) do
      suit = Enum.take(deck, 1)
      {suit, deck}
  end
end

File: com.euchre/src/main/java/com/euchre/Deck.java
package com.euchre;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;

public class Deck {
    private ArrayList<Card> cards;
    public Deck(){
        this.cards = new ArrayList<Card>();
        this.cards.add(new Card("A", "Hearts"));
        this.cards.add(new Card("A", "Diamonds"));
        this.cards.add(new Card("A", "Spades"));
        this.cards.add(new Card("A", "Clubs"));
        this.cards.add(new Card("K", "Hearts"));
        this.cards.add(new Card("K", "Diamonds"));
        this.cards.add(new Card("K", "Spades"));
        this.cards.add(new Card("K", "Clubs"));
        this.cards.add(new Card("Q", "Hearts"));
        this.cards.add(new Card("Q", "Diamonds"));
        this.cards.add(new Card("Q", "Spades"));
        this.cards.add(new Card("Q", "Clubs"));
        this.cards.add(new Card("J", "Hearts"));
        this.cards.add(new Card("J", "Diamonds"));
        this.cards.add(new Card("J", "Spades"));
        this.cards.add(new Card("J", "Clubs"));
        this.cards.add(new Card("10", "Hearts"));
        this.cards.add(new Card("10", "Diamonds"));
        this.cards.add(new Card("10", "Spades"));
        this.cards.add(new Card("10", "Clubs"));
        this.cards.add(new Card("9", "Hearts"));
        this.cards.add(new Card("9", "Diamonds"));
        this.cards.add(new Card("9", "Spades"));
        this.cards.add(new Card("9", "Clubs"));
        this.cards.add(new Card("8", "Hearts"));
        this.cards.add(new Card("8", "Diamonds"));
        this.cards.add(new Card("8", "Spades"));
        this.cards.add(new Card("8", "Clubs"));
    }

    public Deck init(){
        return this;
    }

    public Deck shuffle(){
        Collections.shuffle(this.cards);
        return this;
    }

    public Card deal(){
        Card c = cards.stream().findFirst().get();
        cards.remove(c);
        return c;
    }



}
